src/common/NetworkPrivate.cpp

```diff
@@ -54,360 +149,311 @@ QString NetworkData::getHash()
-                    return NetworkManager::accessManager.deleteResource(
-                        data->request_);
+    if (!cachedFile.exists() || !cachedFile.open(QIODevice::ReadOnly))
+    {
+        // File didn't exist OR File could not be opened
+        loadUncached(std::move(data));
+        return;
+    }
 
-                case NetworkRequestType::Post:
-                    if (data->multiPartPayload_)
-                    {
-                        assert(data->payload_.isNull());
+    // XXX: check if bytes is empty?
+    QByteArray bytes = cachedFile.readAll();
 
-                        return NetworkManager::accessManager.post(
-                            data->request_, data->multiPartPayload_);
-                    }
-                    else
-                    {
-                        return NetworkManager::accessManager.post(
-                            data->request_, data->payload_);
-                    }
-                case NetworkRequestType::Patch:
-                    if (data->multiPartPayload_)
-                    {
-                        assert(data->payload_.isNull());
+    qCDebug(chatterinoHTTP).noquote() << data->typeString() << "[CACHED] 200"
+                                      << data->request_.url().toString();
 
-                        return NetworkManager::accessManager.sendCustomRequest(
-                            data->request_, "PATCH", data->multiPartPayload_);
-                    }
-                    else
+    data->emitSuccess(
+        {NetworkResult::NetworkError::NoError, QVariant(200), bytes});
+    data->emitFinally();
+}
+
+void load(std::shared_ptr<NetworkData> &&data)
+{
+    if (data->cache_)
+    {
+        std::ignore = QtConcurrent::run([data = std::move(data)]() mutable {
+            loadCached(std::move(data));
+        });
+    }
+    else
+    {
+        loadUncached(std::move(data));
+    }
+}
+
+void NetworkData::emitSuccess(NetworkResult &&result)
+{
+    if (!this->onSuccess_)
+    {
+        return;
+    }
+
+#ifdef CHATTERINO_WARN_SLOW_HTTP
+    if (!this->executeConcurrently_)
+    {
+        runCallback(
+            false, [cb = std::move(this->onSuccess_),
+                    result = std::move(result), url = this->request_.url(),
+                    hasCaller = this->hasCaller_, caller = this->caller_]() {
+                if (hasCaller && caller.isNull())
+                {
+                    return;
+                }
+
+                QElapsedTimer timer;
+                timer.start();
+                cb(result);
+                if (timer.elapsed() > SLOW_HTTP_THRESHOLD)
+                {
+                    qCWarning(chatterinoHTTP)
+                        << "Slow HTTP success handler for" << url.toString()
+                        << timer.elapsed()
+                        << "ms (threshold:" << SLOW_HTTP_THRESHOLD << "ms)";
+                }
+            });
+        return;
+    }
+#endif
```

Comment:
Why is this all a separate runCallback call and not just inside the runCallback below?

Reply:
It was there because the handlers should only be checked when running in the GUI thread (as mentioned in the description). But as you don't see the use, I removed it.